pkg/worker/task/sync_eth_nft_transfers.go

```diff
@@ -54,7 +61,62 @@ func (h *SyncETHNFTTransferTaskHandler) Handler(message *workers.Msg) {
 		panic(fmt.Errorf("SyncNFTTranfers: failed to unmarshal args: %w", err))
 	}
 
-	res, err := h.AlchemyAPI.GetNFTTransfers(castedArgs.Blockchain, castedArgs.Network, castedArgs.ContractAddresses, castedArgs.FromBlock, "latest", castedArgs.PageKey, TransferPageSize)
+	collections, err := h.NftCollectionQuery.QueryAllNFTCollections()
+	if err != nil {
+		panic(fmt.Errorf("SyncNFTTranfers: failed to get NFT collections: %w", err))
+	}
+
+	contractIDToCollectionMap := make(map[authgearweb3.ContractID]ethmodel.NFTCollection)
+	for _, collection := range collections {
+
+		contractID := authgearweb3.ContractID{
+			Blockchain:      collection.Blockchain,
+			Network:         collection.Network,
+			ContractAddress: collection.ContractAddress,
+		}
+
+		contractIDToCollectionMap[contractID] = collection
+	}
+
+	validAddresses := make([]string, 0)
+	for _, contractAddress := range castedArgs.ContractAddresses {
+		contractID := authgearweb3.ContractID{
+			Blockchain:      castedArgs.Blockchain,
+			Network:         castedArgs.Network,
+			ContractAddress: contractAddress,
+		}
+
+		collection, ok := contractIDToCollectionMap[contractID]
+		// collection is not being watched
+		if !ok {
+			continue
+		}
+
+		collectionBlockHeight := collection.FromBlockHeight.ToMathBig()
+
+		// Check whether the collection is synced or being synced
+		// Check whether the collection height is larger than the request height
+		// Check whether the task is an on-going task, i.e not from page 1
+		if collectionBlockHeight.Cmp(big.NewInt(0)) > 0 && collectionBlockHeight.Cmp(castedArgs.FromBlock) > 0 && castedArgs.PageKey == "" {
```

Comment:
Would `collectionBlockHeight.Cmp(big.NewInt(0)) > 0 && collectionBlockHeight.Cmp(castedArgs.FromBlock) > 0` suffice already?

Reply:
Since the block height is updated after we get the page 1 of the result, page 2 would be skipped if we don't guard the page key

Ex.
// Page 1 of the request
collection block height = 0
request = startBlock: 0, endBlock: latest, pageKey: ""

// Page 2 of the request
collection block height = 100000
request = startBlock: 0, endBlock: latest, pageKey: "53a55436-5ac9-4209-b607-76284be9b802"